src/main/java/io/cdap/plugin/successfactors/common/util/ExceptionParser.java

```diff
@@ -0,0 +1,173 @@
+   *
+   * @param msg               stage wise error message for example the caller is
+   *                          - failure while testing OData URL
+   *                          - failure while reading metadata
+   *                          - failure while reading total available record count
+   *                          - failure while reading actual data
+   * @param responseContainer {@code SuccessFactorsResponseContainer} contains the http response details after calling
+   *                          SAP SuccessFactors service
+   * @throws SuccessFactorsServiceException in case of any error scenario, it prepares and throw this exception.
+   */
+  public static void checkAndThrowException(String msg, SuccessFactorsResponseContainer responseContainer)
+    throws SuccessFactorsServiceException {
+
+    String failureMessage = msg;
+    SuccessFactorsError error = null;
+
+    if (responseContainer.getHttpStatusCode() == HttpURLConnection.HTTP_UNAUTHORIZED) {
+      failureMessage += ResourceConstants.ROOT_CAUSE_LOG.getMsgForKey()
+        .concat(ResourceConstants.ERR_INVALID_CREDENTIAL.getMsgForKey());
+
+      throw new SuccessFactorsServiceException(failureMessage, responseContainer.getHttpStatusCode());
+    } else if (responseContainer.getHttpStatusCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      failureMessage += ResourceConstants.ROOT_CAUSE_LOG.getMsgForKey()
+        .concat(ResourceConstants.ERR_INVALID_ENTITY_NAME.getMsgForKey());
+
+      throw new SuccessFactorsServiceException(failureMessage, responseContainer.getHttpStatusCode());
+    } else if (responseContainer.getHttpStatusCode() != HttpURLConnection.HTTP_OK) {
+      String rawResponseString = "";
+      InputStream rawStream = responseContainer.getResponseStream();
+      if (rawStream != null) {
+        try (BufferedReader bufferedReader = new BufferedReader(
+          new InputStreamReader(rawStream, StandardCharsets.UTF_8))) {
+
+          rawResponseString = bufferedReader.lines().collect(Collectors.joining(" "));
+          if (rawResponseString.contains("Refer to https")) {
+            rawResponseString = rawResponseString.substring(0, rawResponseString.indexOf("Refer to https"));
+          }
+        } catch (IOException ioe) {
+          // no-ops
+        }
+      }
+
+      LOG.error("HTTP Code: {}", responseContainer.getHttpStatusCode());
+      LOG.error("Detailed error message: {} {}", msg, rawResponseString);
+
+      try {
+        error = GSON.fromJson(rawResponseString, SuccessFactorsError.class);
+      } catch (JsonSyntaxException | JsonIOException je) {
+        // html errors are only found in case of invalid SuccessFactors service namespace
+        if (rawResponseString.startsWith("<html>")) {
+          failureMessage += ResourceConstants.ERR_INVALID_ENTITY_NAME.getMsgForKey();
+          throw new SuccessFactorsServiceException(failureMessage, SuccessFactorsServiceException.INVALID_SERVICE);
```

Comment:
why don't we just use the http error code returned by successfactor api here and couple of places below where we are returning a custom error code?

Reply:
Passed the response HTTP status code instead of the custom error code at this place. 444c10c0add0efdb6799cd258c03b10d1f479f5e

In the other two places: NO_VERSION_FOUND and INVALID_VERSION_FOUND we're specifically checking for the custom response header.